Zend/zend_inheritance.c

```diff
@@ -1146,20 +1146,24 @@ static zend_always_inline inheritance_status do_inheritance_check_on_method_ex(
 		parent = proto;
 	}
 
-	if (!check_only && child->common.prototype != proto && child_zv) {
+	if (!check_only && child->common.prototype != proto) {
 		do {
+			bool copied = false;
 			if (child->common.scope != ce && child->type == ZEND_USER_FUNCTION) {
 				if (ce->ce_flags & ZEND_ACC_INTERFACE) {
 					/* Few parent interfaces contain the same method */
 					break;
-				} else {
+				} else if (child_zv) {
 					/* op_array wasn't duplicated yet */
 					zend_function *new_function = zend_arena_alloc(&CG(arena), sizeof(zend_op_array));
 					memcpy(new_function, child, sizeof(zend_op_array));
 					Z_PTR_P(child_zv) = child = new_function;
+					copied = true;
 				}
 			}
-			child->common.prototype = proto;
+			if (copied || child_scope == ce) {
+				child->common.prototype = proto;
+			}
 		} while (0);
 	}
```

Comment:
This looks wrong to me.

The initial idea of this code was to update ``child->common.prototype`` and duplicate the function entry if it's necessary. After this change I stopped understanding what this code intend to do. I'm also confused by checking ``child->common.scope != ce`` and ``child_scope == ce`` at the same time.

May be the problem should be fixed with the following patch? I'm not completely sure...
```diff
diff --git a/Zend/zend_inheritance.c b/Zend/zend_inheritance.c
index 491d68714b3..ae3956e039c 100644
--- a/Zend/zend_inheritance.c
+++ b/Zend/zend_inheritance.c
@@ -1154,9 +1154,10 @@ static zend_always_inline inheritance_status do_inheritance_check_on_method_ex(
 		parent = proto;
 	}
 
-	if (!check_only && child->common.prototype != proto && child_zv) {
+	if (!check_only && child->common.prototype != proto
+	 && (child_scope == ce || child_zv)) {
 		do {
-			if (child->common.scope != ce && child->type == ZEND_USER_FUNCTION) {
+			if (child_scope != ce && child->type == ZEND_USER_FUNCTION) {
 				if (ce->ce_flags & ZEND_ACC_INTERFACE) {
 					/* Few parent interfaces contain the same method */
 					break;
```